### PR TITLE
samples & tests: Correct main() type

### PR DESCRIPTION
**Pull request by aescolar**
The application main() in Zephyr is defined as having a prototype:
`void main(void)`, as expected by the kernel init (`bg_thread_main`).

So, correct the different samples and tests that were defined
otherwise.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

@andrewboie Regarding the userspace sample, due to my lack of knowledge in this area, I do not know if a different `main()` is expected so I kept that fix in a separate commit. Please tell if it should be done otherwise.

### CODE REVIEW

**Review by pabigot — DISMISSED** [review]
Makes sense to me; I can't see Zephyr ever being a hosted environment where input to or output from main could be useful.  If that does happen it could be controlled by Kconfig.

**Review by nashif — APPROVED** [review]

**Review by pfalcon — CHANGES_REQUESTED** [review]
I'd suggest to not haste with this and reconsider where the problem lies.

> The application main() in Zephyr is defined as having a prototype:
void main(void)

And that's where the culprit may be. Here's just one reference: https://en.cppreference.com/w/c/language/main_function :

> int main (void) { body } | (1) |  
> int main (int argc, char *argv[]) { body } | (2) |   
> /* another implementation-defined signature */ (since C99)

So, based on that, only in C99 it was allowed to have non-standard signature for main(). Here's how GCC treats that:

~~~
$ gcc -Wall main.c 
main.c:3:6: warning: return type of ‘main’ is not ‘int’ [-Wmain]
 void main(void)
      ^~~~
~~~

Further from https://en.cppreference.com/w/c/language/main_function :

> If the return statement is used, the return value is used as the argument to the implicit call to exit()

We have exit(), we have POSIX subsystem in Zephyr, why introduce confusing low-level discrepancy between "native Zephyr" and "POSIX compatible Zephyr"?

> I can't see Zephyr ever being a hosted environment where input to or output from main could be useful.

It's already used like that in @aescolar's own BOARD=native_posix. It allows to accept argv, passed from command line (and that's kinda confusing; and I don't hint that it should be removed, rather perhaps that "native Zephyr" apps should be configurable to accept argv, to be passed via Kconfig var).


Finally, if there's a strong desire to go NIH way regarding simple baseline conventions like signature (or return type) of main(), can as well take inspiration from Microsoft Windows with it WinMain().

**Comment by aescolar** [comment]
@pfalcon Note that calling the zephyr app main function `main()` in Zephyr is really arbitrary. It could have been really called anything else, as it is not what one would normally understand/expect from a C/POSIX `main()`.
Anyhow with this PR I did not want to get into deeper meaning discussions, but just fix what today is an actual undefined behavior bug in this samples, and something which prevents another bugfix from coming in (as it causes the compiler to error out on the signature discrepancy)
The thing is, today, `bg_thread_main()` calls a function which should be `void main(void)`. Most samples and tests do what is expected. This PR just fixes the handful that do not.

> It's already used like that in @aescolar's own BOARD=native_posix. It allows to accept argv,

I think you may have missunderstood that one: `boards/posix/native_posix/main.c:main()` is the real 
C entry point (`main()`) for the zephyr.exe that is compiled.
Any Zephyr app `main()` in native_posix should have the same `void main(void)` signature: it will not get any parameters, and no return value will be expected.

**Comment by aescolar** [comment]
Just fixed also `tests/kernel/fp_sharing/generic/src/main.c` which I had missed

**Comment by pfalcon** [comment]
@aescolar 

> It could have been really called anything else, as it is not what one would normally understand/expect from a C/POSIX main().

Absolutely, it could be either something completely arbitrary, or something which adheres to a well-known de-facto convention, giving least surprise to the users. If you don't think that my concern is valid/important, feel free to reject my -1's.

> I think you may have missunderstood that one:

I'd say it less important whether I misunderstood it, and more important how people understand it (and I try to speak from known cases of that). Here's an example: https://github.com/atigyi/zephyr/commit/bb844cf7ecb81eb998e20f8aeaaf78b072dcbc49#diff-ab5dbba0b0d52d1bbae0da75625c40a5R97 . People build a project against BOARD=native_posix, run it on Linux, pass command-line arguments to it, and ... think they're ported the project to Zephyr.

We can add to that arbitrary main() prototypes, yeah.

**Comment by aescolar** [comment]
> If you don't think that my concern is valid/important

@pfalcon I did not say your concern was not important. If anything I'm saying that it is tangential to this PR. 
Meaning in this PR I'm not redefining how Zephyr app's `main()` prototype should be. I'm just aligning 14 samples/tests out of ~700 into what is already expected.

The concern you raised would need to be handled and discussed in it's own issue/PR (also note it is an API change).

**Comment by pfalcon** [comment]
> I'm just aligning 14 samples/tests out of ~700 into what is already expected.

Let me also clarify my hypothesis: that there has always been bug in Zephyr headers, defining `void main()`. Despite that bug, people made correct usages of `int main()`. Now this patch propagates the bug thru the codebase, instead of fixing the source of it.

**Review by pfalcon — DISMISSED** [review]
Regardless of my previous concern whether Zephyr's main should have its own special type, or one of the common, well-known types, here's another, independent change request:

samples/net/sockets/dumb_http_server, samples/net/sockets/echo,  samples/net/sockets/echo_async, samples/net/sockets/echo_async_select,  samples/net/sockets/http_get,  samples/posix/gettimeofday are intended to be POSIX-compatible (literally, they are expected to build, and are actually built (Makefile.posix) on a POSIX system, e.g. Linux). For them, the change of main() return type is incorrect, as POSIX mandates that return value of main() gets passed to exit()/_exit(), and may lead to warnings (perhaps, as errors) when building on a POSIX system. Also, non-compliant signature for main() should not appear in the compiler input if CONFIG_POSIX_API is defined. Thanks.

**Comment by stephanosio** [comment]
1. `kernel/init.c`, the only caller of `main`, assumed `void main(void)` and @aescolar is only trying to make all implementations align with this, regardless of whether that main definition is correct or not. Having different function definition and declaration is clearly a bug that needs to be fixed and that is what this PR is addressing.

2. `main` here is not necessarily the "main" defined as per C standard (i.e. it could have been called `zephyr_main`, `app_entry`, `NtProcessStartup`, `WinMain` or pretty much anything else). In that sense, I do not think what C standard says really matters in this case.

3. I do not see the point of having `int main(void)` or `int main(int argc, char **argv)`. This is an OS that is intended to run on a microcontroller, not on a complex multi-layer system that comes with BIOS/EFI/ARC machine firmware that can pass parameters to/handle return from OS.

**Comment by andrewboie** [comment]
> > If the return statement is used, the return value is used as the argument to the implicit call to exit()
> We have exit(), we have POSIX subsystem in Zephyr, why introduce confusing low-level discrepancy between "native Zephyr" and "POSIX compatible Zephyr"?

FWIW returning from main() in Zephyr does not make any implicit exit() call, if we added a return value it would simply be ignored, which also could also be considered confusing.

**Comment by andrewboie** [comment]
> i.e. it could have been called zephyr_main, app_entry, NtProcessStartup, WinMain or pretty much anything else

Perhaps this is really what we need to do -- standardize on `void zephyr_main(void)` and leave the existing main() support around but deprecate it.

**Comment by jmgao** [comment]
Note that if main is defined to return an int, application code can be subtly broken between native_posix and everything else, because main is specifically defined to return 0 if no explicit return happens, but zephyr_app_main isn't main, so you'll likely just fall off the end of the function onto whatever happens to be below it.

It seems to me that the best approach might be to `#define` main to zephyr_main or whatever globally, instead of just in native_posix.

**Comment by aescolar** [comment]
Guys, all these discussions are very fine, but quite off-track.
The situation is as follows:
* Today the kernel assumes a signature for the zephyr application `main()`, this assumption has existed for very long, and it is clearly documented.
* 14 samples/tests have a different one, which was not caught before. These are 14 (undefined behavior) bugs.
* This PR addresses these bugs, and nothing else.
* These set of bugs prevent other bug fixes from being applied as they would break CI revealing them.
* @pfalcon would like this API to be changed to something else, and is blocking this PR hoping that will cause the API to change to what he wants. But the API can only be changed thru other means, with proper deprecation, and is unlikely to change to what he wants. Probably a long discussion will be needed. Therefore blocking this PR is not helping anybody.

@andrewboie You are the TSC approved maintainer of the kernel, and therefore the owner of this app-kernel IF. Therefore you have the right to decide if:
a) @pfalcon review should be dismissed as off-topic, and the patch should be eventually merged (subject to other reviews)
OR
b) You consider that his feedback is on-point and that indeed the bug is in the kernel and in all other samples/tests and documentation. In which case we have a much worse problem.
@andrewboie Please tell which way to go.

@andyross , @jukkar , @tbursztyka , @andrewboie , @tgorochowik , @rveerama1 : Please be so kind as to review this PR independently of the discussion above as this PR touches files you wrote, modified or own, and approve, reject or provide feedback. If you do not think you should review this PR, you are welcome to remove yourselves from the list of reviewers.

**Review by tgorochowik — APPROVED** [review]
+1 for the fix here (civetweb sample which I added), another +1 for discussing suggestions about changing the API (as a separate issue/pr)

**Review by jukkar — APPROVED** [review]

**Comment by pfalcon** [comment]
@stephanosio:

> kernel/init.c, the only caller of main, assumed void main(void)

Well, so maybe that's where the bug lies? Or maybe that's all not even a bug at all, maybe it's actually underspecified, main() can return void or int, it's just that the result is (currently) ignored.

> main here is not necessarily the "main" defined as per C standard (i.e. it could have been called zephyr_main, app_entry, NtProcessStartup, WinMain or pretty much anything else

Exactly, as I said above, if there's strong urge to disregard well-known conventions and establish adhoc one, can as well go for an adhoc name.

> I do not see the point of having int main(void) or int main(int argc, char **argv). This is an OS that is intended to run on a microcontroller,

I don't see how the idea to break well-known conventions and establish confusing adhoc ones stems from intention to run on a microcontroller. Surely, you can respect existing standards and conventions, and be kind to human users in that regard, and still run on a microcontroller?

**Comment by pfalcon** [comment]
@aescolar 

> @pfalcon would like this API to be changed to something else, and is blocking this PR hoping that will cause the API to change to what he wants. 

Not exactly, I just point out that samples/net/sockets/dumb_http_server, samples/net/sockets/echo, samples/net/sockets/echo_async, samples/net/sockets/echo_async_select, samples/net/sockets/http_get, samples/posix/gettimeofday are POSIX application, and rightfully use `int main(...)` signature for its main function, as required by ANSI C/POSIX. So, changing them to `void main(...)` would be deterioration, not improvement.

POSIX subsys is a first-class Zephyr citizen, so I kindly ask to augment this solution to take into account its needs too. It seems that previously, we dealt with this issue via underspecification - both `int` and `void` return types were allowed, but the result was ignored. If there's strong desire to have it specified exactly for "native Zephyr" application as `void`, I'd seriously suggest to still leave it underspecified (specifically, don't provide main() prototype) in case CONFIG_POSIX_API is defined. This would allow for both `int main(void)` and `int main(argc, argv)` forms.

The fact that in case of CONFIG_POSIX_API, result of main() isn't passed to exit/_exit is definitely a bug ("not yet implemented" type, there're many such bugs). I'll be happy to keep it in my list and work on it.

**Review by carlescufi — APPROVED** [review]

**Comment by andyross** [comment]
I've been hiding from the bike shed, but it's been pointed out that that's not really part of my job description.  So...

Come on, folks.  There is no good reason to have a function called "main" that doesn't conform to the ISO C language standard.  No, the return value doesn't mean anything.  Yes, we might very well want something better suited to Zephyr.  But if so then we should work up a different patch with an alternative Zephyr entry point and bike shed about that.  This one seems unambiguously correct (if, OK, maybe a little pedantic).

**Review by andyross — APPROVED** [review]

**Comment by pabigot** [comment]
I originally approved this, before I realized it "broke" existing POSIX samples.  So I've withdrawn my approval.

Given that Zephyr's execution environment is a free-standing one it I fully approve of `void main(void)` as the defined entrypoint.  However, at some point it was agreed that Zephyr should also have POSIX compatibility features.  Unless that decision is reversed, I believe @pfalcon's concerns have merit and should be addressed.  ([My review](https://github.com/zephyrproject-rtos/zephyr/pull/21100#pullrequestreview-325190939) suggested a way to support an alternative signature on a per-application basis.)

But more importantly, I believe those concerns were recorded as a "Changes Requested" which blocks the merge.  My understanding is that dismissing another collaborator's Changes Requested is not allowed under Zephyr's current process.

That this has been done in order to get this PR to pass all checks concerns me.

We do have an undocumented informal process of tagging issues with dev-review and discussing them in a weekly Thursday telecon to try to reach a community agreement and convince a nay-sayer to change their mind.  But as far as I know the only legitimate way to dismiss a Changes Requested from a responsive reviewer is to petition TSC.

Perhaps @aescolar can chime in, as somebody who's working hard to improve Zephyr's process.

**Comment by aescolar** [comment]
Hi @pabigot. Process wise: Indeed the ultimate authority for settling any technical matter is the TSC both now and after that new process could be ratified. And indeed the process to dismiss a review thru a maintainer is not yet ratified. 
(sidetracking into the new process) But also note that this new process certainly includes the option for the reviewer, if he disagrees with the maintainer's ruling, to state so and therefore raise the matter to the TSC (I mentioned this to @pfalcon offline. I also mentioned that I would consider discussing this between the parties and the maintainer before escalating it logical). (the whole point of this process is to try to unblock stalemates in a reasonable way (both time and escalation wise), while still providing proper escalation paths in case of disagreements.
(Note that there seems to be an annoying thing about GH, and it is that a notification is not sent out when a review is dismissed).
(sidetrack end)
Even that this process is not formally ratified, I would considered it logical that in case of disagreements between 2 contributors in an area, one would ask the maintainer to rule, and that if there is still a disagreement the offended party would then raise it further. And that is what has been happening here. Please also note that negative reviews have been dismissed before, based each time in what somebody considered to be fairer and in most cases without escalating it to the TSC. Trying to specify a resolution process is certainly trying to improve on this, to ensure people does not feel they have been treated unfairly or have not been listened to.
I have added a DNM tag to this PR, so as to allow @pfalcon to explicitly state if he agrees with the ruling or not, and to avoid the PR being merged accidentally before.
(On a positive note: I guess we can consider this PR a trial of that new process, which will make us all wiser about its merits )
(technical discussion following)

**Comment by aescolar** [comment]
Now sidetracking into the issue of disagreement:

A few notes (I'm sure you are very aware of most of these):
* The samples in question could still compile as POSIX applications. The only thing that worsens now is that if compiled with `-Wmain` a new warning is generated. So as such this does not break these applications. Also note that at least the one of those applications I tried cannot be compiled as a non-Zephyr app now :(, and that even if it did, the lack of an include for `stdlib.h` (for `exit()`) is actually causing worse warnings by default.
* Today Zephyr cannot possibly handle in any way the return value. So there is no functionality lost as such due to this PR. Looking at it positively, this change would stop misleading somebody who reads the code into thinking the returns are handled. The loss in this PR, would be that, if in the future the expected signature for main would be changed and the return value would be handled, then indeed this app returns would suddenly be handled as expected.
* Indeed the kernel specs a `void main(void)` and that [is documented](https://docs.zephyrproject.org/latest/reference/kernel/threads/system_threads.html). But the lack of a prototype in a common header has lead to not detecting some tests and samples did not follow this. This is a (minor) bug. My motivation for fixing this, was that it blocks #21101, which is a solution for #21094 (and will avoid this bugs being introduced in the future)

(bikeshedding starting, and I really was hoping to do this in a separate issue) 
I think that the best solution would be if we have had the zephyr application main be called something else, say
`void zephyr_application_main(void)`; With that signature and everything else as it is today.
And have the POSIX API add a shim which defined something like:
```
void zephyr_application_main(void){
	int ret;
	
	ret = main();
	
	if (ret) {
		exit(ret);
	}
}
```
Now, unfortunately that is not where we are today. (maybe I'm wrong here but) I'd consider the current `void main(void)` to be part of the public API of Zephyr. So getting there would require deprecating the old API for 2 releases. And until that would be done the solution for the POSIX API could not be implemented. It would also take much more work to deprecate the API than this PR does.

So, in my humble, biassed, opinion, this PR does not cause an actual loss of functionality to those samples, but only a pendantic warning being added if they could be compiled standalone. Any proper solution for POSIX seems to require further work including deprecation anyhow. And therefore I would consider it reasonable to accept this PR.

**Comment by pabigot** [comment]
@aescolar Thanks, I think that's a reasonable path forward.  I do note that the primary role of the dev review meeting is to provide a forum for discussing issues like this, which may be more effective than back-and-forth in github comments, and can provide feedback from experienced community members who don't track every PR discussion.

From the technical side: wouldn't adding `CONFIG_ZEPHYR_HOSTED` which switches the declaration between `void main(void)` and `int main (int argc, char *argv[])` be a workable solution? 

```
#ifdef CONFIG_ZEPHYR_HOSTED
        int main(int argc, char *argv[]);
	char * argv[] = { "Zephyr", NULL };

        (void)main(1, argv);
#else
        void main(void);

        main();
#endif	
```

POSIX examples would select that option, which would default to false.  Naively that seems like a potential simple solution, but @pfalcon should confirm that it would work.

**Comment by aescolar** [comment]
@pabigot Well.. if so then I would reuse `CONFIG_POSIX_API` instead of defining a new `CONFIG_ZEPHYR_HOSTED`. But note that that would not solve the issue of not handling the return value, and note that the kernel knows nothing about `exit()`.
Sidenote: I wouldn't be too keen on adding to the core of the kernel ifdefs based on options from other subsystems. It intermingles things a bit too much.

**Comment by pabigot** [comment]
@aescolar `CONFIG_ZEPHYR_HOSTED` is not the same as `CONFIG_POSIX_API`.  I envision a bootloader or supervisor that initiated transfer to Zephyr (possibly including arguments), and provided a way to communicate the result back.  Having a separate name would make it easier to complete this capability when we find out somebody wants it, without breaking existing users so badly.

**Comment by aescolar** [comment]
After having talked with @pfalcon on slack: he is not pleased with the conclusion, but he does not want to escalate it further (to the TSC). Therefore I remove the DNM.

**Review by ioannisg — COMMENTED** [review]
@aescolar it is now confusing (At least) that several main() function have a "return" statement at the end of the function, which seems redundant, although in other occasions, this return is skipped.

**Comment by aescolar** [comment]
> it is now confusing (At least) that several main() function have a "return" statement at the end of the function, which seems redundant, although in other occasions, this return is skipped.

Changed

**Pull request merged**
